backend/core/src/shared/units-transformations.ts

```diff
@@ -32,85 +34,131 @@ const minMaxCurrency = (baseValue: MinMaxCurrency, newValue: number): MinMaxCurr
   }
 }
 
-const bmrHeaders = ["Studio", "1 BR", "2 BR", "3 BR", "4 BR"]
+// Creates data used to display a table of household size/unit size by maximum income per the AMI charts on the units
+// Unit sets can have multiple AMI charts used, in which case the table displays ranges
+const hmiData = (units: Units, maxHouseholdSize: number) => {
+  // Currently, BMR chart is just toggling whether or not the first column shows Household Size or Unit Type
+  const showUnitType = units[0].bmrProgramChart
 
-const hmiData = (units: Units, byUnitType: UnitSummary[], amiPercentages: string[]) => {
-  const bmrProgramChart = units[0].bmrProgramChart
-  // TODO https://github.com/bloom-housing/bloom/issues/872
-  const amiChartItems = units[0].amiChart?.items || []
-  const hmiHeaders = {
-    householdSize: bmrProgramChart ? "t.unitType" : "listings.householdSize",
-  } as AnyDict
-  const amiValues = amiPercentages
-    .map((percent) => {
-      const percentInt = parseInt(percent, 10)
-      return percentInt
+  type ChartAndPercentage = {
+    percentage: number
+    chart: AmiChart
+  }
+
+  // All unique AMI percentages across all units
+  const allPercentages: number[] = [
+    ...new Set(units.map((unit) => unit.amiPercentage).filter((item) => item != null)),
+  ]
+    .map((percent) => parseInt(percent, 10))
+    .sort(function (a, b) {
+      return a - b
     })
-    .sort()
-  let maxHousehold = 0
-  maxHousehold = byUnitType.reduce((maxHousehold, summary) => {
-    return Math.max(maxHousehold, summary.occupancyRange.max)
-  }, maxHousehold)
 
+  // All unique combinations of an AMI percentage and an AMI chart across all units
+  const uniquePercentageChartSet: ChartAndPercentage[] = [
+    ...new Set(
+      units
+        .map((unit) => {
+          return { percentage: parseInt(unit.amiPercentage, 10), chart: unit.amiChart }
+        })
+        .map((item) => JSON.stringify(item))
+    ),
+  ].map((uniqueSetString) => JSON.parse(uniqueSetString))
```

Comment:
I think it will be more efficient to only reference the amiChart id here and not the full chart. If I'm reading this hmiData function correctly, the only place we need to know the actual chart item data is in `findAmiValueInChart`, so if you updated that function to take the chart id, and used that to get the chart data, I think some of this could be simplified. This does mean that ami charts would need to be fetched separately, but that could and probably should be cached, since I doubt they change very frequently. While this probably doesn't really help speed up processing time, it should reduce the amount of memory used, especially when you take into account that we could take this approach with ami charts everywhere and only ever request the chart id along with listing units.

Reply:
This makes sense to me and I can see the memory gains, but currently I don't think we have access to the amiChart id here without other backend changes. The Unit entity only has the amiChart as `AmiChart`. So I think we need the other PR's [changes](https://github.com/bloom-housing/bloom/pull/1575/files#diff-71332228c9550323008f3cb6962ea93aaa4bf15137a6eeae135ee63dde3ac1a3R56) to move forward with that improvement.

Reply:
I think that's fine, since that PR does fetch the ami charts, we can swap that out there.